src/NLog/Internal/ObjectReflectionCache.cs

```diff
@@ -137,13 +137,24 @@ private static bool TryExtractExpandoObject(Type objectType, out ObjectPropertyI
         {
             foreach (var interfaceType in objectType.GetInterfaces())
             {
-                if (interfaceType.IsGenericType() && interfaceType.GetGenericTypeDefinition() == typeof(IDictionary<,>))
+                if (interfaceType.IsGenericType())
                 {
                     if (interfaceType.GetGenericArguments()[0] == typeof(string))
                     {
-                        var dictionaryEnumerator = (IDictionaryEnumerator)Activator.CreateInstance(typeof(DictionaryEnumerator<,>).MakeGenericType(interfaceType.GetGenericArguments()));
-                        propertyInfos = new ObjectPropertyInfos(null, new[] { new FastPropertyLookup(string.Empty, TypeCode.Object, (o, p) => dictionaryEnumerator.GetEnumerator(o)) });
-                        return true;
+                        if (interfaceType.GetGenericTypeDefinition() == typeof(IDictionary<,>))
+                        {
+                            var dictionaryEnumerator = (IDictionaryEnumerator)Activator.CreateInstance(typeof(DictionaryEnumerator<,>).MakeGenericType(interfaceType.GetGenericArguments()));
+                            propertyInfos = new ObjectPropertyInfos(null, new[] { new FastPropertyLookup(string.Empty, TypeCode.Object, (o, p) => dictionaryEnumerator.GetEnumerator(o)) });
+                            return true;
+                        }
+#if NET4_5
+                        if (interfaceType.GetGenericTypeDefinition() == typeof(IReadOnlyDictionary<,>))
+                        {
+                            var dictionaryEnumerator = (IDictionaryEnumerator)Activator.CreateInstance(typeof(DictionaryEnumerator<,>).MakeGenericType(interfaceType.GetGenericArguments()));
+                            propertyInfos = new ObjectPropertyInfos(null, new[] { new FastPropertyLookup(string.Empty, TypeCode.Object, (o, p) => dictionaryEnumerator.GetEnumerator(o)) });
+                            return true;
+                        }
+#endif
```

Comment:
Would it be possible to use reduce the code duplication?

Reply:
Reduced code duplication